### ISSUE TITLE
Paho mqtt client reconnecting from callback loop

### ISSUE THREAD

**Issue by MartinHjelmare** [sbtinstruments/asyncio-mqtt#26]
## TLDR
There are two cases during connection that will result in reconnecting from the event loop, blocking the loop in the process.

## Background
This library, asyncio-mqtt, wraps paho-mqtt and uses its external loop feature to drive paho-mqtt using the asyncio event loop. The asyncio event loop should never block so it's important to not run blocking I/O inside the event loop.

### paho-mqtt-client details
We hook into the event loop by using the `loop.add_reader` method and connect the paho-mqtt callback `loop_read`. This callback handles the incoming packets from the socket, by calling `_packet_read` and in turn `_packet_handle`. The latter directs to different callbacks depending on the command. One of those cases is the `_handle_connack` callback. There are two cases in this callback that will result in calling the `reconnect` method of the paho mqtt client. This method does blocking I/O. So if we would end up in either of those cases, we would block the event loop during the reconnection.

Case 1:
```py
protocol == MQTTv311 and result == CONNACK_REFUSED_PROTOCOL_VERSION
```

Case 2:
```py
protocol == MQTTv311 and result == CONNACK_REFUSED_IDENTIFIER_REJECTED and self._client_id == b''
```

https://github.com/eclipse/paho.mqtt.python/blob/42f0b13001cb39aee97c2b60a3b4807314dfcb4d/src/paho/mqtt/client.py#L2985-L3003

## Example log with forced incorrect MQTT protocol (case 1)

```
2020-11-27 12:04:31,471 INFO (MainThread) [asyncio_mqtt] Starting client example.
2020-11-27 12:04:31,472 DEBUG (MainThread) [asyncio] Using selector: EpollSelector
2020-11-27 12:04:31,480 DEBUG (ThreadPoolExecutor-0_0) [asyncio_mqtt] Sending CONNECT (u0, p0, wr0, wq0, wf0, c1, k60) client_id=b''
2020-11-27 12:04:31,482 DEBUG (MainThread) [asyncio_mqtt] Received CONNACK (0, 1), attempting downgrade to MQTT v3.1.
2020-11-27 12:04:31,482 DEBUG (MainThread) [asyncio_mqtt] Sending CONNECT (u0, p0, wr0, wq0, wf0, c1, k60) client_id=b''
Error "Operation timed out". Reconnecting in 3 seconds.
2020-11-27 12:04:44,490 DEBUG (ThreadPoolExecutor-0_0) [asyncio_mqtt] Sending CONNECT (u0, p0, wr0, wq0, wf0, c1, k60) client_id=b''
2020-11-27 12:04:44,493 DEBUG (MainThread) [asyncio_mqtt] Received CONNACK (0, 1), attempting downgrade to MQTT v3.1.
2020-11-27 12:04:44,494 DEBUG (MainThread) [asyncio_mqtt] Sending CONNECT (u0, p0, wr0, wq0, wf0, c1, k60) client_id=b''
Error "Operation timed out". Reconnecting in 3 seconds.
^C2020-11-27 12:04:55,391 INFO (MainThread) [asyncio_mqtt] Exiting client example.
```

The first `Sending CONNECT` happens in the thread pool. This is good. But after the downgrade the next `Sending CONNECT` happens in the main thread, ie event loop. That's bad.

## Solution?
I haven't found any easy way of resolving this without changes in paho-mqtt. I've been playing with monkey patching the paho-mqtt `reconnect` method, but it doesn't seem clean so I don't really like that. 

The upside is that these two cases only occur during connection and may also be avoided by making sure the correct MQTT protocol is used and setting the `client_id`.

I'm opening this issue, for verification and information, and discussion if there's anything we can do to mitigate this.

**Comment by frederikaalund** [comment]
Thanks for bringing this issue to my attention. Great to see a proper documentation of the issue and a thorough analysis of the underlying problems (down to the troublesome lines of code). :+1: 

> I haven't found any easy way of resolving this without changes in paho-mqtt. I've been playing with monkey patching the paho-mqtt reconnect method, but it doesn't seem clean so I don't really like that.

It's tempting to monkey-patch our way out of this. Like you, I don't think we should do this. It's a maintenance nightmare.

> The upside is that these two cases only occur during connection and may also be avoided by making sure the correct MQTT protocol is used and setting the client_id.

Good point. Basically, we put the responsibility on the users (to ensure that their server and client protocol versions match). This avoids the blocking reconnect. While this "fixes" the issue, I don't think it's user-friendly enough.

---

I propose that we take this issue to the paho-mqtt developers.

Ideally, I'd like them to reconsider this implicit reconnect logic. It basically breaks the promise of non-blocking semantics in their external event loop API. This is a breaking change so I don't think they'll do this in practice. That's just my guess; I don't know them.

Instead, let's propose a new config entry for `paho.mqtt.Client`. Something like `reconnect_on_protocol_failure` that is `True` per default. paho-mqtt only attempts a reconnect in `_handle_connack` when `reconnect_on_protocol_failure is True`. This proposal is a non-breaking change of paho-mqtt. I think the paho devs likes the sound of that (I know that I would :) ).

In turn, we set `reconnect_on_protocol_failure=False` in `asyncio_mqtt.Client`. This avoids the blocking reconnect. Yes, we lose the implicit reconnect semantics but I'd say that's a fair trade. We can always re-implement the implicit reconnect logic directly in asyncio-mqtt in a later version.

What do you think? Thanks again for bringing this issue to my attention. :+1:

**Comment by MartinHjelmare** [comment]
I think your suggestion sounds good.

So when `reconnect_on_protocol_failure=False` is set and case 1 or 2 happens `_handle_connack` just returns `MQTT_ERR_PROTOCOL`?

**Comment by frederikaalund** [comment]
Great. :)

> So when reconnect_on_protocol_failure=False is set and case 1 or 2 happens _handle_connack just returns MQTT_ERR_PROTOCOL?

Yeah, that's what I would do. Maybe the paho devs have another idea. We'll see.

Do you want to create the issue/pr in paho-mqtt or should?

**Comment by MartinHjelmare** [comment]
If you have time, I think you should do it since it was your idea. But I'm happy to help. Let me know.

How would the config option be set, you think? Add a new method on the paho client to set the option, or a new optional parameter part of an existing method?

**Comment by frederikaalund** [comment]
I'll have a look at it. :+1: 

> How would the config option be set, you think? Add a new method on the paho client to set the option, or a new optional parameter part of an existing method?

I haven't had a look at paho in a while. I'll get back to you on that one.

**Comment by frederikaalund** [comment]
I created a pull request: eclipse/paho.mqtt.python#530. Let's wait and see what happens.

**Comment by solarjoe** [comment]
I don't get the full picture yet, but are they mixing blocking and non-blocking in `reconnect()`?

    def reconnect(self):
    ....
        if self._ssl:
            ....
            sock.settimeout(self._keepalive)
        ....
        self._sock.setblocking(0)

**Comment by frederikaalund** [comment]
> I don't get the full picture yet, but are they mixing blocking and non-blocking in reconnect()?

Indeed, `reconnect` may block. This is per design and is okay for us (we simply don't call reconnect directly).

The actual issue is that paho-mqtt may call `reconnect` implicitly in one of the event handlers. In turn, this may block the event loop.

My patch (optionally) disables the implicit calls to `reconnect`. This way, we don't block the event loop.

**Comment by frederikaalund** [comment]
The paho.mqtt authors merged eclipse/paho.mqtt.python#530. :tada: We now have the `reconnect_on_failure` argument (defaults to `True`). We simply need to set this to `False` inside `Client.__init__` to fix this issue. Does anybody want to do a PR? :)

**Comment by MartinHjelmare** [comment]
We need to wait until the next paho-mqtt release, right?

**Comment by frederikaalund** [comment]
Yes, that's a good point. We can have the PR ready, though, so that it's
easy to merge in as soon as paho.mqtt releases a new version. :)

**Comment by MartinHjelmare** [comment]
I can do a PR in August.

**Comment by JensPfeifle** [comment]
FYI, it seems that we can expect the 1.6 release around the end of September according to https://github.com/eclipse/paho.mqtt.python/issues/531#issuecomment-904538413.

**Comment by solarjoe** [comment]
Any plans for a release and a push to pypi?

**Comment by frederikaalund** [comment]
> Any plans for a release and a push to pypi?

I just pushed v0.11.0 to PyPI. :+1:
